vignettes/python-primer.Rmd

```diff
@@ -0,0 +1,920 @@
+equivalents to the object-oriented style of Python `class`'s.
+
+In Python, however, understanding the basics of how `class` objects work
+is requisite knowledge, because `class`'s are how you organize and find
+methods in Python. (In contrast to R's S3 approach, where methods are
+found by dispatching from a generic). Fortunately, the basics of
+`class`'s are accessible.
+
+Don't be intimidated if this is your first exposure to object oriented
+programming. We'll start by building up a simple Python class for
+demonstration purposes.
+
+```{python}
+class MyClass:
+  pass # `pass` means do nothing.
+
+MyClass
+type(MyClass)
+
+instance = MyClass()
+instance
+type(instance)
+```
+
+Like the `def` statement, the `class` statement binds a new callable
+symbol, `MyClass`. First note the strong naming convention, classes are
+typically `CamelCase`, and functions are typically `snake_case`. After
+defining `MyClass`, you can interact with it, and see that it has type
+`'type'`. Calling `MyClass()` creates a new object **instance** of the
+class, which has type `'MyClass'` (ignore the `__main__.` prefix for
+now). The instance prints with its memory address, which is a strong
+hint that it's common to be managing many instances of a class.
+
+In the first example, we defined an empty `class`, but when we inspect
+it we see that it already comes with a bunch of attributes (`dir()` in
+Python is equivalent to `names()` in R):
+
+```{python}
+dir(MyClass)
+```
+
+#### What are all the underscores?
+
+Python typically indicates that something is special by wrapping the
+name in double underscores. A special double-underscore-wrapped token is
+commonly called a "dunder". "Special" is not a technical term, it just
+means that the token invokes a Python language feature. Some dunder
+tokens are merely ways code authors can plug into specific syntactic
+sugars, others are values provided by the interpreter that would be
+otherwise hard to acquire, yet others are for extending language
+interfaces (e.g., the iteration protocol), and finally, a small handful
+of dunders are truly complicated to understand. Fortunately, as an R
+user looking to use some Python features through reticulate, you only
+need to know about a few easy-to-understand dunders.
+
+The most common dunder method you'll encounter when reading Python code
+is `__init__()`. This is a function that is called when the class
+constructor is called, that is, when a class is **instantiated**. It is
+meant to initialize the new class instance. (In very sophisticated code
+bases, you may also encounter classes where `__new__` is also defined,
+this is called before `__init__`).
+
+```{python}
+class MyClass:
+
+  print("MyClass's definition body is being evaluated")
+
+  def __init__(self):
+    print(self, "is initializing")
+
+print("MyClass is finished being created")
+
+instance = MyClass()
+print(instance)
+
+instance2 = MyClass()
+print(instance2)
+```
+
+A few things to note:
```

Comment:
Interesting! I purposely took it a little slow here. I was worried it maybe a little surprising for R users that a `class` code block is executed only once. Otherwise, people might expect something similar to `def` semantics, where the code block is executed on each call. Also, once you know a class definition is executed only once, the fact that anything defined in the class body is shared by all the instances is a realization that naturally follows....

Reply:
hm, seems like I think differently about this ... I mean, a function _definition_ is also executed just once, right?
But maybe it's just that my first programming language really was Java that makes me find it obvious that a class is defined once ...